### PR TITLE
prep work for using timely dataflow with NLL

### PR DESCRIPTION
**Pull request by nikomatsakis**
Two major changes:

**Two-phase borrows are overhauled.** We no longer have two bits per borrow. Instead, we track -- for each borrow -- an (optional) "activation point". Then, for each point P where the borrow is in scope, we check where P falls relative to the activation point. If P is between the reservation point and the activation point, then this is the "reservation" phase of the borrow, else the borrow is considered active. This is simpler and means that the dataflow doesn't have to care about 2-phase at all, at last not yet.

**We no longer support using the MIR borrow checker without NLL.** It is going to be increasingly untenable to support lexical mode as we go forward, I think, and also of increasingly little value. This also exposed a few bugs in NLL mode due to increased testing.

r? @pnkfelix 
cc @bobtwinkles

### CODE REVIEW

**Comment by nikomatsakis** [comment]
NB. Not 100% tested yet =) will fix as needed.

File: src/librustc/infer/error_reporting/mod.rs [hunk 1]
```diff
@@ -303,28 +303,28 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
     ) {
         debug!("report_region_errors(): {} errors to start", errors.len());
 
-        if will_later_be_reported_by_nll && self.tcx.nll() {
+        if will_later_be_reported_by_nll && self.tcx.use_mir() {
```
Comment (pnkfelix) [inline src/librustc/infer/error_reporting/mod.rs]: eek; while I understand collapsing `-Z nll` into `-Z borrowck=mir`, you've chosen the method name "use_mir" to represent this collapsed flag? I would have expected you to either reuse "nll" again, or to include something about borrow-checking in the method name... "use_mir" just seems very broad and easy to misinterpret...

Comment (nikomatsakis) [reply src/librustc/infer/error_reporting/mod.rs]: indeed I should rename it =)

**Comment by TimNN** [comment]
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/364604871) ([raw log](https://api.travis-ci.org/v3/job/364604871/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
Receiving objects: 100% (329/329), 44.56 KiB | 22.28 MiB/s, done.
---
Resolving deltas: 100% (293/293), completed with 158 local objects.
---
[00:00:54] configure: rust.quiet-tests     := True
---
[00:15:12] error[E0599]: no method named `use_mir_borrowck` found for type `rustc::session::config::BorrowckMode` in the current scope
[00:15:12]   --> librustc_mir/util/borrowck_errors.rs:50:33
[00:15:12]    |
[00:15:12] 50 |             Origin::Mir => mode.use_mir_borrowck(),
---
[00:15:15]   process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustc --crate-name rustc_mir librustc_mir/lib.rs --color always --error-format json --crate-type dylib --emit=dep-info,link -C prefer-dynamic -C opt-level=2 -C metadata=97e644ee6c469e67 -C extra-filename=-97e644ee6c469e67 --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/release/deps --extern rustc_apfloat=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_apfloat-1299638d641ea770.rlib --extern bitflags=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libbitflags-9866e194db82a141.rlib --extern graphviz=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libgraphviz-0bde40de32995f14.so --extern arena=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libarena-f2778bd6cd1c5bdd.so --extern log=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/liblog-3e35efb9e5bc7a9a.rlib --extern rustc_errors=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_errors-609e2421d03f9c9a.so --extern rustc_const_math=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_const_math-895f7ddc4467bb8d.so --extern log_settings=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/liblog_settings-70f7ed359bd54256.rlib --extern serialize=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libserialize-c04ded78717d5d67.so --extern serialize=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libserialize-c04ded78717d5d67.rlib --extern rustc_data_structures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_data_structures-93cb1ddd29ab61a4.so --extern rustc_back=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_back-861dba9fe03aa669.so --extern syntax=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libsyntax-d3b6fcf798f7d22a.so --extern syntax_pos=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libsyntax_pos-9f3518d56a01456f.so --extern rustc=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc-23f5dcecdda8feb4.so --extern byteorder=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libbyteorder-ca61dfec7c40c4d4.rlib -L native=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/build/backtrace-sys-8b35e3c2ea935fab/out/.libs -L native=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/build/miniz-sys-63734d0048644b22/out` (exit code: 101)
[00:15:15] warning: build failed, waiting for other jobs to finish...
[00:17:16] error: build failed
[00:17:16] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "--target" "x86_64-unknown-linux-gnu" "--release" "--locked" "--color" "always" "--features" " jemalloc" "--manifest-path" "/checkout/src/rustc/Cargo.toml" "--message-format" "json"
[00:17:16] expected success, got: exit code: 101
[00:17:16] thread 'main' panicked at 'cargo must succeed', bootstrap/compile.rs:1085:9
[00:17:16] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:17:16] travis_fold:end:stage0-rustc
[00:17:16] travis_time:end:stage0-rustc:start=1523366597370910721,finish=1523367265133942713,duration=667763031992
[00:17:16] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap build
[00:17:16] Build completed unsuccessfully in 0:11:21
[00:17:16] make: *** [all] Error 1
[00:17:16] Makefile:28: recipe for target 'all' failed
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

**Comment by nikomatsakis** [comment]
@pnkfelix ok addressed your nit :)

File: src/test/compile-fail/E0501.rs [hunk 2]
```diff
@@ -26,10 +26,12 @@ fn foo(a: &mut i32) {
         inside_closure(a)
     };
     outside_closure_1(a); //[ast]~ ERROR cannot borrow `*a` as mutable because previous closure requires unique access
-                         //[mir]~^ ERROR cannot borrow `*a` as mutable because previous closure requires unique access
```
Comment (pnkfelix) [inline src/test/compile-fail/E0501.rs]: Of course, there was a reason that these comments were indented so much: To allow ease of comparison with the line above.

Having said that, we all hope that compare-mode=nll will provide a much better path forward, so I don't really care to try to preserve details like this here.

**Comment by pnkfelix** [comment]
This all seems fine.

If I were writing this I might have changed the `-Z borrowck=compare` code to use the new suffix "(Nll)" in place of the old "(Mir)", just to emphasize the broader scope of the two things being compared now... but then again, given that `-Z borrowck=mir` now *implies* NLL, maybe its for the best that it be denoted by "(Mir)" in that context.

**Comment by nikomatsakis** [comment]
Rebased. That was a bit of a tricky rebase (git diffs were pretty messed up). =) But I think I got it right.

**Pull request merged**